### ISSUE TITLE
Unable to use aspectj in java library

### ISSUE THREAD

**Issue by auke-** [bazelbuild/bazel#4597]
### Description of the problem / feature request:

When a project uses aspectj (aspectjtools-1.8.13.jar) in a java_library ijar fails with the error 'Bad magic efbfbdef'

When the ijar utility is used separately the output is:
...
INFO: StripClass: org/aspectj/org/eclipse/jdt/internal/compiler/parser/UpdateParserFiles.class
Bad magic efbfbdef

It seems that the ijar utility isn't able to read the UpdateParserFiles class file properly or it contains an invalid magic number.

Due to this issue we aren't able to migrate from Maven to Bazel.

#### Checkout bazel project and build ijar
git clone "https://github.com/bazelbuild/bazel.git"
cd bazel
bazel build //third_party/ijar

#### Download aspjectj jar
mvn org.apache.maven.plugins:maven-dependency-plugin:3.0.2:get -Dartifact=org.aspectj:aspectjtools:1.8.13

#### Run ijar against aspectj jar
./bazel-bin/third_party/ijar/ijar -v ~/.m2/repository/org/aspectj/aspectjtools/1.8.13/aspectjtools-1.8.13.jar

### What operating system are you running Bazel on?

MacOS Sierra

### What's the output of `bazel info release`?

release 0.10.0

###  Have you found anything relevant by searching the web?

When searching for 0xEFBFBDEF i found this stackoverflow question regarding encoding errors:
https://stackoverflow.com/questions/5421879/objectinputstream-happy-with-fileinputstream-not-happy-with-getresourceasstream

**Comment by cushon** [comment]
This is not a problem with `ijar`, it's a problem with `org/aspectj/org/eclipse/jdt/internal/compiler/parser/UpdateParserFiles.class`. class files are required to start with `0xcafebabe`, and it does not. `javap` can't read that file either:

```
wget http://central.maven.org/maven2/org/aspectj/aspectjtools/1.8.13/aspectjtools-1.8.13.jar
javap -fullversion -cp aspectjtools-1.8.13.jar org/aspectj/org/eclipse/jdt/internal/compiler/parser/UpdateParserFiles
10-ea+37
Error: error while reading constant pool for org/aspectj/org/eclipse/jdt/internal/compiler/parser/UpdateParserFiles: unexpected tag at #1: 191
```

There's a related bug report here: https://bugs.eclipse.org/bugs/show_bug.cgi?id=510978

**Comment by auke-** [comment]
Ah, I see. Since the bug report you refer to is more than a year old what would be a workaround?
- make ijar emit a warning instead of an error (what would be te consequence for Bazel?)
- remove the incorrect class file from the jar (would that be possible using a genrule?)
- ...?

Completely removing aspectj from our project isn't an option at this time.

**Comment by cushon** [comment]
> make ijar emit a warning instead of an error

We don't want ijar to silently tolerate invalid class files. This error indicates something is wrong with the dependency, and the best solution is to fix the dependency.

If fixing aspectj isn't feasible, and deleting the class file is enough to get your build working, using a genrule as you suggested sounds like a fine work-around.

I don't think there's anything to do here on the Bazel side.

**Comment by auke-** [comment]
I understand the best solution is to fix the error upstream, but since that isn't likely to happen any time soon i'm looking for alternatives.

The genrule approach has some major downsides, like the fact that i'm not able to use the default maven_jar rule including authentication and the reference to the source jars.

The proposed solution doesn't silently ignore invalid class files, it emits a message and uses the whole file instead of a stripped down version. How is this any different from what happens when an invalid constant is encountered (see https://github.com/bazelbuild/bazel/blob/master/third_party/ijar/classfile.cc#L1520 and https://github.com/bazelbuild/bazel/blob/master/third_party/ijar/classfile.cc#L1558)?

**Comment by cushon** [comment]
You can still use `maven_jar` - the files can be accessed using e.g. `@aspectjtools//jar:file` or skylark (see [`JavaInfo`](https://docs.bazel.build/versions/master/skylark/lib/JavaInfo.html)), and then repackaged for with the source jars using `java_import` or [`java_common.create_provider`](https://docs.bazel.build/versions/master/skylark/lib/java_common.html#create_provider).

The handling of unrecognized constant pool entries is different because those might still be well-formed class files, they just aren't ones ijar knows how to read. (e.g. it doesn't handle some of the JDK 9 constrant pool entries yet.) Ideally we would fix that and do more validation, not less.

**Comment by auke-** [comment]
Ok, thanks. I'll try to use maven_jar from a Skylark rule.

**Comment by auke-** [comment]
If I only was a little less impatient... Aspectj released [version 1.9.0](https://mvnrepository.com/artifact/org.aspectj/aspectjtools/1.9.0) which doesn't include the corrupt class file.
